src/connectionbenchmark/benchmarktasktransfer.cpp

```diff
@@ -18,46 +19,61 @@ constexpr const char* MULLVAD_DEFAULT_DNS = "10.64.0.1";
 #endif
 
 namespace {
-Logger logger(LOG_MAIN, "BenchmarkTaskDownload");
+Logger logger(LOG_MAIN, "BenchmarkTaskTransfer");
 }
 
-BenchmarkTaskDownload::BenchmarkTaskDownload(const QUrl& url)
-    : BenchmarkTask(Constants::BENCHMARK_MAX_DURATION_DOWNLOAD),
+BenchmarkTaskTransfer::BenchmarkTaskTransfer(BenchmarkType type,
+                                             const QUrl& url)
+    : BenchmarkTask(Constants::BENCHMARK_MAX_DURATION_TRANSFER),
+      m_type(type),
       m_dnsLookup(QDnsLookup::A, url.host()),
-      m_fileUrl(url) {
-  MVPN_COUNT_CTOR(BenchmarkTaskDownload);
+      m_url(url) {
+  MVPN_COUNT_CTOR(BenchmarkTaskTransfer);
 
   connect(this, &BenchmarkTask::stateChanged, this,
-          &BenchmarkTaskDownload::handleState);
+          &BenchmarkTaskTransfer::handleState);
   connect(&m_dnsLookup, &QDnsLookup::finished, this,
-          &BenchmarkTaskDownload::dnsLookupFinished);
+          &BenchmarkTaskTransfer::dnsLookupFinished);
 }
 
-BenchmarkTaskDownload::~BenchmarkTaskDownload() {
-  MVPN_COUNT_DTOR(BenchmarkTaskDownload);
+BenchmarkTaskTransfer::~BenchmarkTaskTransfer() {
+  MVPN_COUNT_DTOR(BenchmarkTaskTransfer);
 }
 
-void BenchmarkTaskDownload::handleState(BenchmarkTask::State state) {
+void BenchmarkTaskTransfer::handleState(BenchmarkTask::State state) {
   logger.debug() << "Handle state" << state;
 
   if (state == BenchmarkTask::StateActive) {
-    // Start DNS resolution
+    if (m_type == BenchmarkDownload) {
+// Start DNS resolution
 #if !defined(MVPN_DUMMY)
-    m_dnsLookup.setNameserver(QHostAddress(MULLVAD_DEFAULT_DNS));
+      m_dnsLookup.setNameserver(QHostAddress(MULLVAD_DEFAULT_DNS));
 #endif
 
 #if QT_VERSION >= 0x060400
-#  error Check if QT added support for QDnsLookup::lookup() on Android
+#     error Check if QT added support for QDnsLookup::lookup() on Android
 #endif
 
 #ifdef MVPN_ANDROID
-    NetworkRequest* request =
-        NetworkRequest::createForGetUrl(this, m_fileUrl.toString());
-    connectNetworkRequest(request);
-    m_elapsedTimer.start();
+      NetworkRequest* request =
+          NetworkRequest::createForGetUrl(this, m_fileUrl.toString());
+      connectNetworkRequest(request);
+      m_elapsedTimer.start();
 #else
-    m_dnsLookup.lookup();
+      m_dnsLookup.lookup();
 #endif
+    } else if (m_type == BenchmarkUpload) {
+      UploadDataGenerator* uploadData =
+          new UploadDataGenerator(Constants::BENCHMARK_MAX_BITS_UPLOAD);
+      uploadData->open(UploadDataGenerator::ReadOnly);
+
+      // TODO: Create multiple network requests for upload
+      NetworkRequest* request = NetworkRequest::createForUploadData(this,
+          m_url.toString(), uploadData);
+      connectNetworkRequest(request);
```

Comment:
This part is just for testing purposes and as soon as we have endpoints we can post data to this can look very similar to the `BenchmarkDownload` network requests above.